src/Marten/Internal/Sessions/EventTracingConnectionLifetime.cs

```diff
@@ -0,0 +1,286 @@
+using System;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Marten.Services;
+using Npgsql;
+
+namespace Marten.Internal.Sessions;
+
+internal class EventTracingConnectionLifetime :
+    IConnectionLifetime
+{
+    private const string MartenCommandExecutionStarted = "MartenCommandExecutionStarted";
+    private const string MartenCommandExecutionFailed = "MartenCommandExecutionFailed";
+    private const string MartenBatchExecutionStarted = "MartenBatchExecutionStarted";
+    private const string MartenBatchExecutionFailed = "MartenBatchExecutionFailed";
+    private const string MartenBatchPagesExecutionStarted = "MartenBatchPagesExecutionStarted";
+    private const string MartenBatchPagesExecutionFailed = "MartenBatchPagesExecutionFailed";
+
+    private readonly IConnectionLifetime _innerConnectionLifetime;
+    private readonly OpenTelemetryOptions _openTelemetryOptions;
+    private readonly Activity _databaseActivity;
+
+    public EventTracingConnectionLifetime(IConnectionLifetime innerConnectionLifetime, OpenTelemetryOptions openTelemetryOptions, Activity databaseActivity)
+    {
+        if (innerConnectionLifetime == null)
+        {
+            throw new ArgumentNullException(nameof(innerConnectionLifetime));
+        }
+
+        if (openTelemetryOptions == null)
+        {
+            throw new ArgumentNullException(nameof(openTelemetryOptions));
+        }
+
+        if (databaseActivity == null)
+        {
+            throw new ArgumentNullException(nameof(databaseActivity));
+        }
+
+        Logger = innerConnectionLifetime.Logger;
+        CommandTimeout = innerConnectionLifetime.CommandTimeout;
+        _innerConnectionLifetime = innerConnectionLifetime;
+        _openTelemetryOptions = openTelemetryOptions;
+        _databaseActivity = databaseActivity;
+    }
+
+    public ValueTask DisposeAsync()
+    {
+        return _innerConnectionLifetime.DisposeAsync();
+    }
+
+    public void Dispose()
+    {
+        _innerConnectionLifetime.Dispose();
+    }
+
+    public IMartenSessionLogger Logger { get; set; }
+    public int CommandTimeout { get; }
+    public int Execute(NpgsqlCommand cmd)
+    {
+        if (_openTelemetryOptions.TrackConnectionEvents)
+        {
+            _databaseActivity.AddEvent(new ActivityEvent(MartenCommandExecutionStarted));
+        }
+
+        try
+        {
+            return _innerConnectionLifetime.Execute(cmd);
+        }
+        catch (Exception e)
+        {
+            if (_openTelemetryOptions.TrackConnectionEvents)
+            {
+                _databaseActivity.AddEvent(new ActivityEvent(MartenCommandExecutionFailed, DateTimeOffset.UtcNow, new ActivityTagsCollection(new KeyValuePair<string, object>[] {new KeyValuePair<string, object>("ExceptionType", e.GetType()) })));
```

Comment:
There's an extension method `RecordException` (from OTel) that can be used here.

Reply:
That would require taking a hard dependency on OTEL. This implementation is also missing the stack trace.